### PR TITLE
BL-1104: Out-of-memory errors cropping large images

### PR DESCRIPTION
**Pull request by phillip-hopper**

### CODE REVIEW

File: PalasoUIWindowsForms/ImageToolbox/ImageToolboxControl.cs [hunk 1]
```diff
@@ -163,24 +158,27 @@ private ListViewItem AddControl(string label, Bitmap bitmap, string imageKey, Sy
...
 
-			if (_inIndexChanging)
+			if (_previousSelectedIndex == _toolListView.SelectedIndices[0])
 				return;
 
 			try
 			{
-				_inIndexChanging = true;
-
-				if (_toolListView.SelectedItems.Count == 0)
```
Comment (papeh) [inline PalasoUIWindowsForms/ImageToolbox/ImageToolboxControl.cs]: if `SelectedItems.Count == 0` is still a possibility (e.g. when there are no items available to select), an exception will be thrown outside the try block.

File: PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs [hunk 2]
```diff
@@ -217,45 +242,23 @@ private void ImageCropper_Resize(object sender, EventArgs e)
...
 
-		private float GetImageToCanvasScaleFactor()
+		private float GetImageToCanvasScaleFactor(Image img)
 		{
 			var availArea = new Rectangle(BorderSize, BorderSize, Width - (2*BorderSize), Height - (2*BorderSize));
-			float hProportion = (float) availArea.Width/((float) _originalImage.Width);
-			float vProportion = (float) availArea.Height/((float) _originalImage.Height);
+			float hProportion = (float)availArea.Width / ((float)img.Width);
+			float vProportion = (float)availArea.Height / ((float)img.Height);
 			return Math.Min(hProportion, vProportion);
```
Comment (papeh) [inline PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs]: the above method no longer relies on instance data so could be made static.

File: PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs [hunk 3]
```diff
@@ -143,7 +144,32 @@ public PalasoImage Image
...
 				//other code changes the image of this palaso image, at which time the PI disposes of its copy,
 				//so we better keep our own.
-				_originalImage = (Image) value.Image.Clone();
+
+				// save the original in a temp file instead of an Image object to free up memory
+				_savedOriginalImage = TempFile.CreateAndGetPathButDontMakeTheFile();
+				value.Image.Save(_savedOriginalImage.Path, ImageFormat.Png);
+				
+				// make a reasonable sized copy to crop
+				if (value.Image.Width > 1000)
```
Comment (papeh) [inline PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs]: could check height, too.

File: PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs [hunk 4]
```diff
@@ -358,33 +359,26 @@ public Image GetCroppedImage()
...
-				CheckForInvalidImage();
-
 				//NB: this worked for tiff and png, but would crash with Out Of Memory for jpegs.
 				//This may be because I closed the stream? THe doc says you have to keep that stream open.
 				//Also, note that this method, too, lost our jpeg encoding:
 				//          return bmp.Clone(selection, _image.PixelFormat);
 				//So now, I first copy it, then clone with the bounds of our crop:
 
-				using (var copy = new Bitmap(_originalImage)) //**** here we lose the jpeg rawimageformat, if it's a jpeg. Grrr.
+				using (var copy = new Bitmap(_savedOriginalImage.Path)) //**** here we lose the jpeg rawimageformat, if it's a jpeg. Grrr.
```
Comment (JohnThomson) [inline PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs]: So the original code ended up copying the original twice, plus the output which might be nearly as big. Both versions have to make a bitmap of the whole original as well as the new bitmap. If the original was a very compressed jpg or sparse line drawing, we may still be using a lot of memory we don't need.
The top answer at http://stackoverflow.com/questions/734930/how-to-crop-an-image-using-c has another approach to copying part of an image into a new bitmap that might be worth considering.

**Comment by JohnThomson** [comment]
One suggestion to consider...otherwise LGTM

File: PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs (hunk 2 above)
Comment (phillip-hopper) [inline PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs]: Line 253 uses the Width and Height properties of the UserControl.

File: PalasoUIWindowsForms/ImageToolbox/ImageToolboxControl.cs (hunk 1 above)
Comment (phillip-hopper) [inline PalasoUIWindowsForms/ImageToolbox/ImageToolboxControl.cs]: Code moved to inside the try-catch block.

File: PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs (hunk 3 above)
Comment (phillip-hopper) [inline PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs]: Now checking both height and width.

File: PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs (hunk 4 above)
Comment (phillip-hopper) [inline PalasoUIWindowsForms/ImageToolbox/Cropping/ImageCropper.cs]: I couldn't determine if one method is better than the others. It looks like they all call GDI+ to do the actual copying.

**Comment by papeh** [comment]
:+1:

**Pull request merged**